# 01 Importing libraries

In [29]:
# Importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 02 Importing Dataframes

In [31]:
# Importing customer dataframe
path = r'C:\Users\vacgo\Documents\Achievement 4\Instacart Basket Analysis'
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [32]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_latest_20241115_2.pkl'))

In [33]:
# Set prices greater than 100 to NaN
ords_prods_merge.loc[ords_prods_merge['prices'] > 100, 'prices'] = np.nan

In [34]:
print(ords_prods_merge.columns)

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_from_last_order', 'new_customer',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'merge_2', 'price_range_loc',
       'busiest_day', 'busiest_days_new', 'busiest_period_of_day', 'max_order',
       'loyalty_flag', 'ave_spent', 'spender', 'median_days_to_order',
       'frequent_cust'],
      dtype='object')


# 03 Check dataframe for customers

In [36]:
customers.shape

(206209, 10)

In [37]:
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [38]:
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [39]:
# Assign user_id as string
customers['user_id'] = customers['user_id'].astype('str')

In [40]:
# Rename columns
customers_new = customers.rename(columns = {'First Name': 'first_name', 'Surnam': 'last_name', 'Gender' : 'gender', 'STATE': 'state', 'Age' : 'age', 'n_dependants':'dependants'}, inplace = False)

In [41]:
customers_new.head()

user_id first_name last_name  gender       state  age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   dependants fam_status  income  
0           3    married  165665  
1           0     single   59285  
2           2    married   99568  
3           0     single   42049  
4           1    married   40374

In [42]:
# Check date_joined values
customers_new['date_joined'].sample(30)

150437     5/17/2019
66747      1/20/2018
196976      2/9/2020
87100      5/16/2018
148086      5/3/2019
70868      2/12/2018
188028    12/19/2019
154328      6/8/2019
123315    12/12/2018
89063      5/27/2018
70773      2/11/2018
193034     1/17/2020
14434      3/24/2017
126480    12/30/2018
148343      5/5/2019
144760     4/14/2019
167796     8/25/2019
12365      3/12/2017
36298      7/30/2017
100         1/1/2017
162701     7/28/2019
2807       1/17/2017
19721      4/24/2017
117130     11/6/2018
148101      5/3/2019
167158     8/21/2019
103368     8/18/2018
14568      3/25/2017
178703    10/26/2019
3381       1/20/2017
Name: date_joined, dtype: object

#### Seems okay

In [44]:
# Check state names if it is 50
customers_new['state'].nunique()

51

In [45]:
customers_new['state'].value_counts().sort_index()

state
Alabama                 4044
Alaska                  4044
Arizona                 4044
Arkansas                4044
California              4044
Colorado                4044
Connecticut             4044
Delaware                4044
District of Columbia    4044
Florida                 4044
Georgia                 4044
Hawaii                  4044
Idaho                   4044
Illinois                4044
Indiana                 4044
Iowa                    4044
Kansas                  4043
Kentucky                4043
Louisiana               4043
Maine                   4043
Maryland                4043
Massachusetts           4043
Michigan                4043
Minnesota               4043
Mississippi             4043
Missouri                4043
Montana                 4043
Nebraska                4043
Nevada                  4043
New Hampshire           4043
New Jersey              4043
New Mexico              4043
New York                4043
North Carolina          4043
North Da

#### state column seems fine

In [47]:
# Check for null values
customers_new.isnull().sum()

user_id            0
first_name     11259
last_name          0
gender             0
state              0
age                0
date_joined        0
dependants         0
fam_status         0
income             0
dtype: int64

In [48]:
customers_no_name = customers_new[customers_new['first_name'].isnull() == True]

In [49]:
customers_no_name

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        dependants fam_status  income  
53               2    married   41709  
73               0     single   82518  
82               3    married  117099  
99               1    married  155673  
105              3    married  151819  
...            ...        ...     ...  
206038           3    married   87783  
206044           2    married  108488  
206090           1    married   45275  
206162           0     single   56325  
206171           2    married   59222  

[11259 rows x 10 columns]

#### There are 11,259 out of 206,209 entries with no first names. Which is around 5% of the data. Might still include it in the analysis if the first name is not important in the analysis

In [51]:
# Dropping first_name and last_name columns in the dataframe due to privacy
customers_clean = customers_new.drop(['first_name', 'last_name'], axis = 1)

In [52]:
# Check for duplicates
customers_dups = customers_new[customers_new.duplicated()]

In [53]:
customers_dups.shape

(0, 10)

#### No duplicate values

In [55]:
# Checking for mixed type in columns of customers_new
#for col in customers_new.columns.tolist():
#    if len(customers_new[col].apply(type).unique())>1:
#        print(f"Mixed-type data found in column: {col}")
#    else:
#        print(f"No mixed-type data in column: {col}")

In [56]:
# Assign first_name as string - if necessary
# customers_new['first_name'] = customers_new['first_name'].astype('str')

In [57]:
# Checking for mixed type in columns in customers_clean
for col in customers_clean.columns.tolist():
    if len(customers_clean[col].apply(type).unique())>1:
        print(f"Mixed-type data found in column: {col}")
    else:
        print(f"No mixed-type data in column: {col}")

No mixed-type data in column: user_id
No mixed-type data in column: gender
No mixed-type data in column: state
No mixed-type data in column: age
No mixed-type data in column: date_joined
No mixed-type data in column: dependants
No mixed-type data in column: fam_status
No mixed-type data in column: income


In [58]:
customers_clean.dtypes

user_id        object
gender         object
state          object
age             int64
date_joined    object
dependants      int64
fam_status     object
income          int64
dtype: object

In [59]:
ords_prods_merge.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_from_last_order      float64
new_customer                 bool
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
merge_2                  category
price_range_loc            object
busiest_day                object
busiest_days_new           object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
ave_spent                 float64
spender                    object
median_days_to_order      float64
frequent_cust              object
dtype: object

In [60]:
# Assign user_id as string in the ords_prods_merge
ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype('str')

In [61]:
ords_prods_merge.dtypes

order_id                    int64
user_id                    object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_from_last_order      float64
new_customer                 bool
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
merge_2                  category
price_range_loc            object
busiest_day                object
busiest_days_new           object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
ave_spent                 float64
spender                    object
median_days_to_order      float64
frequent_cust              object
dtype: object

In [62]:
# Counting unique user_id for customers
customers_new['user_id'].nunique()

206209

In [63]:
# Counting unique user_id for ords_prods_merge
ords_prods_merge['user_id'].nunique()

206209

# 04 Combining the two dataframes

In [65]:
# Combine the customers_clean into ords_prods_merge
orders_products_all = ords_prods_merge.merge(customers_clean, on='user_id')


In [66]:
orders_products_all.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2539329       1             1                  2                  8   
2   2539329       1             1                  2                  8   
3   2539329       1             1                  2                  8   
4   2539329       1             1                  2                  8   

   days_from_last_order  new_customer  product_id  add_to_cart_order  \
0                   NaN          True         196                  1   
1                   NaN          True       14084                  2   
2                   NaN          True       12427                  3   
3                   NaN          True       26088                  4   
4                   NaN          True       26405                  5   

   reordered  ...      spender  median_days_to_order          frequent_cust  \
0          0  ...  Low spender                  20.5  Non-frequent customer   
1          0  ...  Low spender                  20.5  Non-frequent customer   
2          0  ...  Low spender                  20.5  Non-frequent customer   
3          0  ...  Low spender                  20.5  Non-frequent customer   
4          0  ...  Low spender                  20.5  Non-frequent customer   

   gender    state age date_joined dependants fam_status  income  
0  Female  Alabama  31   2/17/2019          3    married   40423  
1  Female  Alabama  31   2/17/2019          3    married   40423  
2  Female  Alabama  31   2/17/2019          3    married   40423  
3  Female  Alabama  31   2/17/2019          3    married   40423  
4  Female  Alabama  31   2/17/2019          3    married   40423  

[5 rows x 32 columns]

In [67]:
orders_products_all.shape

(32404859, 32)

In [68]:
# Check for null values
#orders_products_all.isnull().sum()

In [69]:
#ords_prods_merge.isnull().sum()

# 05 Exporting

In [71]:
orders_products_all.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_all_20241118.pkl'))